homeassistant/components/sql/sensor.py

```diff
@@ -70,34 +73,54 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
         unit = query.get(CONF_UNIT_OF_MEASUREMENT)
         value_template = query.get(CONF_VALUE_TEMPLATE)
         column_name = query.get(CONF_COLUMN_NAME)
+        query_template = query.get(CONF_QUERY_TEMPLATE)
 
+        if query_str and query_template:
+            raise Exception("Both query and query_template are defined. Choose one.")
         if value_template is not None:
             value_template.hass = hass
-
+        if query_template is not None:
+            query_template.hass = hass
         sensor = SQLSensor(
-            name, sessmaker, query_str, column_name, unit, value_template
+            hass,
+            name,
+            sessmaker,
+            query_str,
+            column_name,
+            unit,
+            value_template,
+            query_template,
         )
         queries.append(sensor)
+    async_add_entities(queries)
 
-    add_entities(queries, True)
+    return True
 
 
 class SQLSensor(Entity):
     """Representation of an SQL sensor."""
 
-    def __init__(self, name, sessmaker, query, column, unit, value_template):
+    def __init__(
+        self, hass, name, sessmaker, query, column, unit, value_template, query_template
+    ):
         """Initialize the SQL sensor."""
+        self.hass = hass
         self._name = name
-        if "LIMIT" in query:
-            self._query = query
+        self._query_template = query_template
+
+        if query is not None:
+            if "LIMIT" in query:
+                self._query = query
+            else:
+                self._query = query.replace(";", " LIMIT 1;")
         else:
-            self._query = query.replace(";", " LIMIT 1;")
+            self._query = None
```

Comment:
We should make sure that the query return a single value

Reply:
If "query" is in the config, it does return a single value with the logic above.